### PR TITLE
add SymbolTable for Block

### PR DESCRIPTION
**Pull request by Superjom**
This is one part of Block's Implementation according to [compile period of block design](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/block.md#the-compilation-of-blocks).

The block's implementation will be split into several stages, this is the first one which supports block's compile, to enable InferShape's development to be done parallel.

resolves: https://github.com/PaddlePaddle/Paddle/issues/4171

### CODE REVIEW

File: paddle/framework/block.cc [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+   See the License for the specific language governing permissions and
+   limitations under the License. */
+
+#include "paddle/framework/block.h"
+#include "paddle/platform/enforce.h"
+
+namespace paddle {
+namespace framework {
+
+const OpDesc* SymbolTable::NewOp() {
```
Comment (jacquesqiao) [inline paddle/framework/block.cc]: does not really new an Op?

Comment (Superjom) [reply paddle/framework/block.cc]: New a Op Desc, SymbolTable is a container of compile-time instances, this op means an op in compile-time, that is a OpDesc.

File: paddle/framework/block.cc [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+const VarDesc* SymbolTable::FindVar(const std::string& name,
+                                    bool recursive) const {
+  auto it = vars_.find(name);
+  if (it != vars_.end()) {
+    return &(it->second);
+  }
+  if (recursive) {
+    return parent_->FindVar(name, true);
```
Comment (jacquesqiao) [inline paddle/framework/block.cc]: should judge if parent_ is null

File: paddle/framework/block.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+  /*
+   * Create a BlockDesc based on the registered VarDesc and OpDesc.
+   */
+  BlockDesc Compile() const;
+
+ private:
+  SymbolTable* parent_{nullptr};
+  std::vector<OpDesc> ops_;
+  std::map<std::string, VarDesc> vars_;
```
Comment (reyoung) [inline paddle/framework/block.h]: unordered_map

File: paddle/framework/block.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+namespace paddle {
+namespace framework {
+
+/*
+ * SymbolTable is the container of OpDesc and VarDesc.
+ */
+class SymbolTable {
+ public:
+  SymbolTable() {}
+  explicit SymbolTable(SymbolTable* parent) : parent_(parent) {}
```
Comment (reyoung) [inline paddle/framework/block.h]: Where is deconstructor, should we disable the `Copy & Assign`?

**Comment by wangkuiyi** [comment]
It seems that due to the merge of https://github.com/PaddlePaddle/Paddle/pull/4241, this PR could be closed? @Superjom

**Pull request closed without merging**